### DISCUSSION TITLE
New OpenAI confusion

### DISCUSSION THREAD

**Discussion by mhyben** [oobabooga/text-generation-webui#4765] [Q&A]
Hi, I am struggling with the new OpenAI API.  

I start the server with the following command:
python3 server.py --api --listen

The server is started with no problem except that IP address it reports it is running on is http://0.0.0.0. Of course no such address is found but I can find it running on the machine IP address on a port that the server reported:
"Running on local URL:  http://0.0.0.0:7860"

When I open that address in a browser, I can control the GUI with no problem (download and load models and also use them via chat). The problem arise when I try to access it via API.

This is my python script (copied from wiki [Open AI API section](https://github.com/oobabooga/text-generation-webui/wiki/12-%E2%80%90-OpenAI-API)):

    import requests
    
    url = "http://<machine IP address>:7860/v1/chat/completions"
    
    headers = {
        "Content-Type": "application/json"
    }
    
    history = []
    
    while True:
        user_message = input("> ")
        history.append({"role": "user", "content": user_message})
        data = {
            "mode": "chat",
            "character": "user",
            "messages": history
        }
    
        response = requests.post(url, headers=headers, json=data, verify=False)
        if response.status_code == 200:
            assistant_message = response.json()['choices'][0]['message']['content']
            history.append({"role": "assistant", "content": assistant_message})
            print(assistant_message)
        else:
            print(f'Error code: {response.status_code} {response.reason}')
       

I get the 404 Not found error. I tried multiple combinations of url suffixes:
/v1/
/v1/completions
/v1/chat
/api/v1/generate
etc.
.. but nothing seems to work.

I also experimented with the command to run the server:
python3 server.py --public-api --listen
python3 server.py --api
etc.

I just cannot make API running. What am I doing wrong?
Thanks.

### System info
Pop Os Linux 64-bit

## Edit:
I was able to resolve the problem by specifying the api_port in the command for starting the server.
So the command looked like this:
python3 server.py --api --extension openai chat --api-port 1234 --listen

And then the code looked like this:

    url = f'http://{HOST_IP}:1234/v1/chat/completions'
    url_list = f'http://{HOST}/v1/models'

headers = {
        "Content-Type": "application/json"
    }

    def __init__(self):
        self.reset_history()

    def reset_history(self):
        self.history = []

    def run(self, prompt: str):
        self.history.append({"role": "user", "content": prompt})
        data = {'prompt': prompt,
                "mode": "chat",
                "character": "Example",
                "messages": self.history,

                "max_tokens": 8,
                "temperature": 1,
                "top_p": 0.9,
                "seed": 10
                }

        response = requests.post(self.url, headers=self.headers, json=data, verify=False)

        if response.status_code == 200:
            message = response.json()['choices'][0]['message']['content']
            self.reset_history()
            return message
        else:
            print(f'Error code: {response.status_code}')

if __name__ == '__main__':
    ui = WebUI()
    ui.list_models()
    while True:
        print(ui.run(input()))

**Comment by zencyon** [comment]
you have :
```
history.append({"role": "user", "content": user_message})
data = {
        "mode": "chat",
        "character": "user",
        "messages": history
    }
```
the 'character' should be set to the name of one of your character files in webui/characters folder, and not "user"

it should be:
```
history.append({"role": "user", "content": user_message})
data = {
        "mode": "chat",
        "character": "Example",
        "messages": history
    }
```
